cmp/form/FormModel.js

```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * Reset fields to initial values and reset validation.
+     */
+    reset() {
+        this.eachMember(m => m.reset());
+    }
+
+    /**
+     * Set the initial value of all fields, and reset form.
+     *
+     * If initial values are undefined for any field, existing initial values
+     * will be used.
+     *
+     * This is the main entry point for loading new data (or an empty new record)
+     * into the form for editing.
+     */
+    init(initialValues = {}) {
+        this.eachMember(m => m.init(initialValues[m.name]));
+    }
+
+    //--------------------------
+    // Validation
+    //---------------------------
+    /** @member {ValidationState} - the current validation state. */
+    @computed
+    get validationState() {
+        const VS = ValidationState,
+            states = this.mapMembers(m => m.validationState);
+        if (states.includes(VS.NotValid)) return VS.NotValid;
+        if (states.includes(VS.Unknown)) return VS.Unknown;
+        return VS.Valid;
+    }
+
+    /**
+     * True if any of the fields contained in this model are in the process
+     * of recomputing their validation state.
+     */
+    @computed
+    get isValidationPending() {
+        return this.someMember(m => m.isValidationPending);
+    }
+
+    /** True if all fields are valid. */
+    get isValid() {
+        return this.validationState == ValidationState.Valid;
+    }
+
+    /** True if any fields are not valid. */
+    get isNotValid() {
+        return this.validationState == ValidationState.NotValid;
+    }
+
+    /**
+     * Return a resolved validation state of the form, waiting for any pending
+     * validations to complete, if necessary.
+     *
+     * @param {Object} [c]
+     * @param {boolean] [c.display] - true to activate validation display
+     *      for the form after validation has been peformed.
+     *
+     * @returns {Promise<ValidationState>}
+     */
+    async validateAsync({display = true} = {}) {
+        const promises = this.mapMembers(m => m.validateAsync({display}));
+        await Promise.all(promises);
+        return this.validationState;
+    }
+
+    /**
+     * Activate Display of all fields.
+     */
+    displayValidation() {
+        this.eachMember(m => m.displayValidation());
+    }
+
+    //----------------------------
+    // Dirty State
+    //----------------------------
+    /**
+     * True if any fields have been changed since last reset/initialization.
+     */
+    isDirty() {
+        return this.someMember(m => m.isDirty);
+    }
+
+    //---------------------------
+    // Implementation
+    //---------------------------
+    eachMember(fn) {
+        this.fields.forEach(fn);
+        this.children.forEach(fn);
+    }
+
+    mapMembers(fn) {
+        return [
+            ...this.fields.map(fn),
+            ...this.children.map(fn)
+        ];
+    }
+
+    someMember(fn) {
+        return this.fields.some(fn) || this.children.some(fn);
+    }
+
```

Comment:
Rather than these `xxxMember` methods, wouldn't it be more straightforward to have a getter than returns the combined array of members and then directly operate on that wherever these are used?

e.g.
```
get members() {
  return [...this.fields, ...this.children];
}

this.members.forEach()
this.members.map()
this.members.some()
```